### ISSUE TITLE
CodeQL C++ extractor does not recognize Clang's `-nostdlibinc` flag

### ISSUE THREAD

**Issue by abcSup** [github/codeql#16767]
CodeQL C++ extractor does not recognize Clang's `-nostdlibinc` flag.

When cross-compiling a target for arm64 with Bionic in AOSP, Android (Soong) build system [adds the `-nostdlibinc` flag](https://cs.android.com/android/platform/superproject/main/+/main:build/soong/cc/config/global.go;drc=ffebc5b71f811b4d04a32115b636837a5483f6c0;l=183) when invoking `clang`/`clang++` to prevent it from searching for header files in the host standard library directories.

However, CodeQL C++ extractor does not recognize the `-nostdlibinc` and throws a warning when encountering it:

```
$ /path/to/codeql/cpp/tools/linux64/extractor --mimic '/path/to/aosp-main/prebuilts/clang/host/linux-x86/clang-r510928/bin/clang++' -nostdlibinc A.cpp
[E 23:00:13 2711401] CodeQL C/C++ Extractor 2.17.4                                                                                                                                                                                                                                                                           
[E 23:00:13 2711401] CODEQL_EXTRACTOR_CPP_SCRATCH_DIR does not exist; not saving separate log file.                                                                                                                                                                                                                          
[E 23:00:13 2711401] Could not open log file; not saving separate log file.                                                                                                                                                                                                                                                  
[E 23:00:13 2711401] Current directory: /usr/local/google/home/zifantan/debug-trace-command                                                                                                                                                                                                                                  
[E 23:00:13 2711401] Command: /path/to/codeql/cpp/tools/linux64/extractor --mimic '/path/to/aosp-main/prebuilts/clang/host/linux-x86/clang-r510928/bin/clang++' -nostdlibinc A.cpp                                                                            
Warning: Could not open file './compiler_mimic_cache/67d147be206e'                                                                                                                                                                                                                                                           
Warning: Unrecognised command line argument -nostdlibinc
...
[E 23:00:13 2711401] Processed command line: /path/to/codeql/cpp/tools/linux64/extractor <...> '-I/usr/include/v1/c++' -I/path/to/aosp-main/prebuilts/clang/host/linux-x86/clang-r510928/lib/clang/18/include -I/usr/local/include -I/usr/include -- A.cpp
```

## Expected

The Clang `-nostdlibinc` flag is defined in clang's [doc](https://github.com/llvm/llvm-project/blob/93ffe1792fd9a985b96fee1105b399b5196a15bc/clang/docs/CommandGuide/clang.rst?plain=1#L678) as follows:
> Do not search the standard system directories for include files, but do
  search compiler builtin include directories.

I believe setting the `-nostdlibinc` flag should be similar to setting the `-nostdinc++` flags while keeping the compiler builtin include directories.

It is expected that CodeQL C++ extractor does not include the C and C++ host standard library headers (such as `-I/usr/include/v1/c++`, `-I/usr/local/include` and `-I/usr/include`) in the processed command line.


## More details
First, I found that some source files are not extracted correctly in the database generated for a AOSP arm64 target.

After looking into the extractor logs, I found the C++ extractor failed to process some compiler invocations because there are some undefined identifiers errors.

```
[E 19:01:47 150254] Warning[extractor-c++]: In construct_text_message: "system/logging/liblog/include/android/log.h", line 204: error: identifier "size_t" is undefined
...
[E 19:01:49 150254] Warning[extractor-c++]: In main: Extractor exiting with code 1
```

Going through the log, especially the cflags and cxxflags, I noticed that the extractor added the `'-I/usr/include/c++/v1'`, which causes it to resolve the `stddef.h` to the host standard system headers (`/usr/lib/llvm-16/include/c++/v1/stddef.h`) instead of the AOSP's compiler builtin include directories (`/path/to/aosp-main/prebuilts/clang/host/linux-x86/clang-r510928/lib/clang/18/include/stddef.h`).

```
[E 19:01:45 150254] Processed command line: /path/to/codeql-home/codeql/cpp/tools/linux64/extractor <...> '-I/usr/include/c++/v1' -I/path/to/aosp-main/prebuilts/clang/host/linux-x86/clang-r5109
28/lib/clang/18/include -I/usr/local/include -I/usr/include -- frameworks/base/libs/hwui/jni/BitmapFactory.cpp
...
[E 19:01:47 150254] Already archived /path/to/codeql-dbs/aosp/src/usr/lib/llvm-16/include/c++/v1/stddef.h
```

Full extractor log for compiling a single `frameworks/base/libs/hwui/jni/BitmapFactory.cpp`: [extractor.log](https://github.com/user-attachments/files/15844811/extractor.log)

Then, after some analysis by comparing with a successful host build, I found CodeQL C++ extractor does not recognize the `-nostdlibinc` correctly, so it includes the host standard library headers. Consequently, this causes the undefined `size_t` identifier error because it is expected to be correctly defined by the compiler builtin `stddef.h` header.

**Comment by mbg** [comment]
Hi @abcSup 👋 

Thanks for reporting this. I will pass it on to the relevant engineering team to look into

**Comment by jketema** [comment]
Hi @abcSup ,

We've fixed the issue. The fix should be part of CodeQL 2.18.0, which will be released in a few weeks time.

**Comment by abcSup** [comment]
Hi @mbg and @jketema, thank you so much for the quick fix!
